packages/cli/src/commands/tunnel/index.ts

```
@@ -52,55 +51,83 @@ const tunnel: CommandModule<unknown, TunnelCommandArgs> = {
+          createLogtoResponseHandler({
+            proxyResponse,
+            request,
+            response,
+            logtoEndpointUrl,
+            tunnelServiceUrl,
+            verbose,
+          })
+      );
+      const proxyExperienceServerRequest = conditional(url && createProxy(url));
+      const proxyExperienceStaticFileRequest = conditional(path && createStaticFileProxy(path));
+
+      const server = http.createServer((request, response) => {
+        consoleLog.info(`[${chalk.green(request.method)}] ${request.url}`);
+
+        // Tunneling the requests to Logto endpoint
+        if (isLogtoRequestPath(request.url)) {
+          void proxyLogtoRequest(request, response);
+          return;
+        }
+
+        if (proxyExperienceServerRequest) {
+          void proxyExperienceServerRequest(request, response);
+          return;
+        }
+
+        if (proxyExperienceStaticFileRequest) {
+          void proxyExperienceStaticFileRequest(request, response);
+        }
+      });
+
+      server.listen(port, () => {
+        const serviceUrl = new URL(`http://localhost:${port}`);
+        consoleLog.info(
+          `🎉 Logto tunnel service is running!
+  ${chalk.green('➜')} Your custom sign-in UI is hosted on: ${chalk.blue(serviceUrl.href)}
+
+  ${chalk.green('➜')} Don't forget to update your app's endpoint:
+
+      ${chalk.gray('From:')} ${chalk.bold(endpoint)}
+      ${chalk.gray('To:')}   ${chalk.bold(serviceUrl.href)}
+
+  ${chalk.green('➜')} ${chalk.gray(`Press ${chalk.white('Ctrl+C')} to stop the tunnel service.`)}
+  ${chalk.green('➜')} ${chalk.gray(
+    `Use ${chalk.white('-v')} or ${chalk.white('--verbose')} to print verbose output.`
+  )}
+          `
+        );
+      });
+
+      server.on('error', (error: Error) => {
+        if ('code' in error && error.code === 'EADDRINUSE') {
+          consoleLog.error(`Port ${port} is already in use, trying another one...`);
+          startServer(port + 1);
```

Or we can throw an error and terminate the problem, to keep it simple

---

usually when no port is specified this can be implemented

---

OK, I think this is a more popular practice recently.
Let's move on.

---

but it will be executed anyway. i'm good with it though